R/rowWilcoxonTests.R

```diff
@@ -72,26 +71,84 @@
 #' cls_perm <- replicate(B, sample(cls))
 #' system.time(fwt <- rowWilcoxonTests(mat, categ = cls_perm, alternative = "two.sided"))
 #' 
-rowWilcoxonTests <- function(mat, categ, 
-                              alternative = c("two.sided", "less", "greater"), 
-                              correct = TRUE) {
+
+rowWilcoxonTests <- function (mat, categ, alternative = c("two.sided", "less", "greater"), 
+                                correct = TRUE) 
+{
+    alternative <- match.arg(alternative)
+    stopifnot(all(categ %in% c(0, 1)))
+    categ <- as.matrix(categ)
+    levels(categ) <- NULL
+    apply(categ, 2, sansSouci:::categCheck, n = ncol(mat))
+    B <- ncol(categ)
+    m <- nrow(mat)
+    n_obs <- nrow(categ)
+    
+    rks <- rowRanks(mat, ties.method = "average")
+    nx <- colSums(categ)
+    ny <- n_obs - nx
+    
+    min_stat <- nx * (nx + 1)/2 
+    
+    stats <- rks %*% categ
+    # stats <- stats - min_stat 
+    stats <- sweep(stats, MARGIN = 2, STATS = min_stat, FUN = "-") # t(t(stats) - min_stat) 
+    
+    rks2 <- rks 
+    mode(rks2) <- "integer"
+    n_ties <- rowTabulates(rks2) 
+    ties <- rowSums(n_ties^3 - n_ties) 
+    # sigma <- sqrt((nx * ny/12) * ((ny + nx + 1) - ties/((ny + nx) * (ny + nx - 1))))
+    sigma <- t(sqrt((nx * ny/12) * ((ny + nx + 1) - matrix(rep(ties, B), nrow = B, byrow = T)/((ny + nx) * (ny + nx - 1)))))
```

Comment:
This is probably not optimal: the formula is of the form t(a - M/b), where a and b are vectors and M is a matrix. So it would be safer to have the "-" and "/" replaced by 'sweep' operations. I leave it to you to decide whether you want to change this now.